Damage Pipeline Update

Caltinor opened the pull request:
This PR aims to create a more cohesive chain of events when entities are dealt damage.

For an in depth explanation of the need for this PR, please see [This writeup by Tslat](https://gist.github.com/Tslat/0ef11e57036f7fe253268b5a64aa455e).

## Overview
this PR adds two (2) new events, renames four (4), an override method for damage, and adds a protected field (`damageContainer`) to `LivingEntity` to track damage throughout the damage sequence.  

### `DamageContainer` Interface
This new field in `LivingEntity` tracks everything that happens during the sequence and is passed to each event.  This means that for a single attack, no detail about the sequence from a prior step is obscured or inaccessible.  For example, in later events, the modified and original damage are referenceable.  Additionally values such as shield blocked damage, armor reduction, enchant reduction, absorption reduction, and mob effect reduction are all accessible (provided they are set at that point in the sequence).   Since this container is provided to each event, it is also where modifications take place so that they persist between events.  

When the sequence is complete, the damage container is discarded until a new attack is received.  This is also a protected method to prevent access from the outside, since its sole purpose is to be a transient container for event-driven modifications happening internally.

### The Sequence at a High Level
The "damage sequence" starts when the entity's `hurt` method is invoked and ends when damage is applied to the entity health.  the sequence is as follows
1. invocation of `hurt`
3. invulnerability check. sequence ends if invulnerable   <= `EntityInvulnerabilityCheckEvent` (new)
4. cancel if on client, dead or dying, or damage is fire and entity is fire resistant
5. first NeoForge hook to cancel or modify damage.  <= `LivingIncomingDamageEvent` (formerly `LivingAttackEvent`)
6. shield/blocking check and damage to shield  <= `LivingShieldBlockEvent` (formerly `ShieldBlockEvent`)
7. freezing check
8. invocation of `actuallyHurt`
10. apply armor reduction  <= `ArmorHurtEvent` (new)
11. apply magic reduction
12. apply absorption
13. NeoForge hook to alter damage and cancel if zero <= `LivingDamageEvent.Pre` (formerly `LivingDamageEvent`)
14. entity health reduced by damage
15. after-changes event to capture final state  <= `LivingDamageEvent.Post` (formerly `LivingHurtEvent`)


### Invulnerability Check
Currently the vanilla invulnerability check prevents modders from having any decision power on the receiving end of damage if vanilla decides the entity is invulnerable.  The new event provides the following benefits:
1. Modders always have a way to listen to an entity being attacked, even if the damage is nothing
2. they can now override invulnerability
3. There is now an earlier point to cancel damage.

### `LivingIncomingDamageEvent` (formerly `LivingAttackEvent`)
Renamed to indicate this is the first stage in the damage sequence.  This event is where most of the change should happen.  Through `DamageContainer#addModifier` armor, enchant, mob effect, and absorption values are changed through the provided callbacks.  Changing the damage in this event will have the most downstream effects.

###  Shield changes and the new `LivingShieldBlockEvent` (formerly `ShieldBlockEvent`)
The current `ShieldBlockEvent` is nested inside vanilla checks as to whether the entity is blocking.  Like the invulnerability check, this was decision power taken from modders.  The new event is now part of the condition, so the damage blocked, the damage to the shield, and whether the player is considered blocking are all handled by the event.

### `ArmorHurtEvent` (new) and Reductions from Armor, Magic, and Absorption
There is currently no way to modify damage to armor durabililty except to modify the damage to the entity through two different events.  This event gives explicit control over the armor durability damage portion.  Furthermore, the modifiers for damage reduction from armor, enchantments, mob effects, and absorption are applied in the `DamageContainer` instance.  

### `LivingDamageEvent.Pre` (formerly `LivingDamageEvent`)
Renamed to indicate this is damage incoming to the entity and cannot be interrupted.  The event is no longer cancellable for this reason.  However, setting the damage to zero does still terminate the damage sequence.  At this stage, all reductions have taken place and have been applied to the `newDamage` value.  This is the last chance for users to change the final damage to the entity, but it will not revert and armor or shield durability losses.

### `LivingDamageEvent.Post` (formerly `LivingHurtEvent`)
Renamed to indicate this shows changes already applied to the entity.  This event is not cancellable, and receives an immutable version of the DamageContainer.  This event works more as a trigger for events that need to fire when an entity takes damage AND needs to know the amount of damage received.

### added `ILivingEntityExtension#onDamageTaken`
a default noop method that executes after the damage sequence has finished and allows custom entities a means to apply custom behavior during `actuallyHurt` without having to override the method.

Caltinor committed:
Damage Pipeline Changes

Caltinor committed:
tweaks to implementation

- function-based modifiers
- interface for damage container
- javadocs

Caltinor committed:
add internal annotation to internal class

Caltinor commented:
Updated with input from discussions on discord.  Original post has been updated to reflect these changes as well.

KnightMiner commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,304 @@
+    /**
+     * Adds a callback modifier to the vanilla damage reductions. Each function will be performed in sequence
+     * on the vanilla value at the time the {@link Reduction} type is set by vanilla.
+     * <h4>Note: only the {@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent IncomingDamageEvent}
```
Isn't this supposed to say `EntityPreDamageEvent`?

Caltinor replied:
yes.  fixed.

KnightMiner commented on src/main/java/net/neoforged/neoforge/event/entity/living/ArmorHurtEvent.java:
```diff
@@ -0,0 +1,72 @@
+public class ArmorHurtEvent extends PlayerEvent implements ICancellableEvent {
+    private final Map<EquipmentSlot, ItemStack> armorItemStack;
+    private final Map<EquipmentSlot, Float> originalDamage;
+    private final Map<EquipmentSlot, Float> damage;
```
Any reason to not use 1 map containing an object with the stack, original damage, and new damage?

Caltinor replied:
Would there be any tangible performance gains from dropping the overhead of 3 `EnumMap`s (the instantiations use EnumMaps) and replacing the values with a class that contains the values?  It couldn't be a record since we expect `damage` to be changed.   (edit: I changed the types in this class to better show they are EnumMaps for reading clarity)

KnightMiner replied:
Mostly just makes it easier to iterate, rather than needing to loop through the slots and call 3 different getters and a setter.

Caltinor replied:
updated.

KnightMiner commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageBlockEvent.java:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * Controls if {@link LivingEntity#hurtCurrentlyUsedShield} is called.
+     * 
+     * @return If the shield item will take durability damage or not.
```
This java doc sounds like its describing a boolean return, not a float return. Same with the setter. Its also a little weird the block amount has a boolean and float while the shield damage is just a float, any reason its inconsistent?

Its also unclear to me how the default works, at a glance shields don't take damage at all unless you tell it to in the event? Shouldn't the default be they take damage if original blocked is true?

Caltinor replied:
This used to be a boolean when this was `ShieldBlockEvent`.  updated.

KnightMiner replied:
Its still unclear to me why you use a boolean for whether it should block, but a non-zero float for shield damage. Why not non-zero floats for both? Why not booleans with both?

The only thing I see is you could potentially damage the shield without blocking any damage? Which I doubt anyone would use.

My concern is its more effort to just make a shield block a certain amount of damage, as you always have to remember to update how much damage the shield is taking. One possibility is making shield damage a 2 state thing, by default (say, NaN or a boolean), the shield damage will return the damage blocked, but if you call the shield damage setter it overrides that and uses the value passed into the setter (which can be zero)

Caltinor committed:
javadoc and type clarity changes

Caltinor committed:
update armor event

Caltinor committed:
resolve missed formatting error

KnightMiner commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -250,14 +262,31 @@ public static boolean onLivingUseTotem(LivingEntity entity, DamageSource damageS
+        ArmorHurtEvent event = NeoForge.EVENT_BUS.post(new ArmorHurtEvent(armorMap, player));
+        if (event.isCanceled()) return;
+        event.getArmorMap().forEach((slot, entry) -> {
+            if ((!source.is(DamageTypeTags.IS_FIRE) || !entry.armorItemStack.getItem().isFireResistant()) && entry.armorItemStack.getItem() instanceof ArmorItem) {
```
shouldn't this be done in the loop before the event instead of after? For non-armor items, either don't include it in `armorMap` at all, or include it with 0 as long as its not empty (so people can damage non-armor items).

For fire damage, just set it to 0 if the armor stack is fire resistance and its fire damage, but let the event override that behavior if they wish.

Caltinor replied:
I like that change, yeah.  i'll add that now.

KnightMiner commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -250,14 +262,31 @@ public static boolean onLivingUseTotem(LivingEntity entity, DamageSource damageS
+        if (event.isCanceled()) return;
+        event.getArmorMap().forEach((slot, entry) -> {
+            if ((!source.is(DamageTypeTags.IS_FIRE) || !entry.armorItemStack.getItem().isFireResistant()) && entry.armorItemStack.getItem() instanceof ArmorItem) {
+                Float finalDamage = event.isCanceled() ? entry.originalDamage : entry.newDamage;
```
Why does canceling mean we deal the original damage? Shouldn't canceling mean we deal 0 damage? Generally canceling a forge event means that behavior does not happen at all, mods or vanilla. I cannot see a benefit to a mod being able to say "ignore all other mods and use vanilla logic". I can however see a benefit to saying "your armor does not take damage because the damage is being redirected to another source like forge energy"

Caltinor replied:
technically the armor wouldn't be damaged at all if the event was cancelled.  So the only scenario that would actually harm the armor is the `newDamage`.  I confusing redundancy that I just fixed.

Caltinor committed:
armor hook reorg

KnightMiner commented on patches/net/minecraft/world/entity/player/Inventory.java.patch:
```diff
@@ -48,3 +48,18 @@
+                 if ((!p_150073_.is(DamageTypeTags.IS_FIRE) || !itemstack.getItem().isFireResistant()) && itemstack.getItem() instanceof ArmorItem) {
+                     itemstack.hurtAndBreak(
+-                        (int)p_150074_, this.player, p_35997_ -> p_35997_.broadcastBreakEvent(EquipmentSlot.byTypeAndIndex(EquipmentSlot.Type.ARMOR, i))
++                            (int)p_150074_, this.player, p_35997_ -> p_35997_.broadcastBreakEvent(EquipmentSlot.byTypeAndIndex(EquipmentSlot.Type.ARMOR, i))
```
I think these two lines in the patch are just an indentation change, probably accidental change? If there is something else different here, would be more clear without the indentation change

Caltinor replied:
removing the indentation and repatching doesn't seem to create any changes that git picks up on.  I even deleted the patch file drastically changed the indentation and regenerated the patches, and this still shows up.

KnightMiner commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageBlockEvent.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    /**
+     * Set if the shield will take durability damage or not.
```
This setter doc still reads like a boolean, think you just fixed it on the getter.

Caltinor replied:
I updated all the javadocs to capture both the new functionality of the methods and the new functionality of the event as a whole

Caltinor committed:
minor tweaks - int cast, shield damage

Caltinor committed:
shield damage value verification

Caltinor committed:
update javadoc, max block amount, og dmg amount

Caltinor committed:
javadoc update to refresh upstream tasks

Shadows-of-Fire commented:
Going to tentatively set this as targeting 1.21, due to the scope of other changes that are currently slated for 1.20.5.  If we find ourselves in an accelerated timeline before the 20.5 stable release, we can revisit this decision.

Caltinor committed:
fix 1.20.5 merge errors.

Caltinor committed:
Fix branch updating errors

Caltinor committed:
fix porting errors

Caltinor committed:
resolve gametest issues

Shadows-of-Fire commented on patches/net/minecraft/world/entity/Entity.java.patch:
```diff
@@ -292,6 +292,19 @@
+             || this.invulnerable && !p_20122_.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !p_20122_.isCreativePlayer()
+             || p_20122_.is(DamageTypeTags.IS_FIRE) && this.fireImmune()
+             || p_20122_.is(DamageTypeTags.IS_FALL) && this.getType().is(EntityTypeTags.FALL_DAMAGE_IMMUNE);
++        return net.neoforged.neoforge.common.CommonHooks.onEntityInvulnerablityCheck(this, p_20122_, isVanillaInvulnerable);
```
Attempting to fire the event here is risky, but firing it at the call sites is untenable.  At the least, I don't think we can reasonably provide the `isVanillaInvulerable` field since it is not necessarily correct - the common case of overrides to this method check additional conditions and then `&&` them with `super.isInvulnerableTo`.

Caltinor replied:
The only overrides where this occurs is:
- Warden: checks if digging animation is active and damage source bypasses immunity
- Ghast: checks entity invulnerability, same damage source check, and if the damage is not a deflected fireball
- Breeze: damage source is in the breeze immune tag or source is from a breeze (self included)

I am of the opinion that the immunity bypass tags are compatible with this feature mainly because they are configurable.      A user or modder still has control over the tag contents to handle immunity if it matters for these 3 mobs.  The other aspects relate to reasonable invulnerabilities (breeze not affecting itself, warden (de)spawning).  The Ghast fireball is the only other one that might be worth patching.  I agree that patching each of these mobs for the sake of uniformity would be untenable in the long run though.

> I don't think we can reasonably provide the isVanillaInvulerable field since it is not necessarily correct
Perhaps rename?  This way it's clearer that it isn't absolute in its scope.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -9,6 +9,14 @@
+     protected Brain<?> brain;
+     private boolean skipDropExperience;
+     protected float appliedScale = 1.0F;
++    protected net.neoforged.neoforge.common.damagesource.DamageContainer.InternalDamageContainer damageContainer;
```
Missing `@Nullable` annotation.  This should also include a javadoc comment about when this field is instantiated and when it returns to being null again.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -132,36 +140,48 @@
+ 
+             this.noActionTime = 0;
+-            float f = p_21017_;
++            float f = p_21017_ = this.damageContainer.getNewDamage();
```
This patch can be changed to the addition of `p_21017_ = this.damageContainer.getNewDamage();` above the `float f = p_21017_;` line instead of patching it out.  It should also include a `// Neo: ` comment stating why we override the parameter instead of just the local.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -132,36 +140,48 @@
-+                if(!ev.isCanceled()) {
-+                if(ev.shieldTakesDamage()) this.hurtCurrentlyUsedShield(p_21017_);
++            net.neoforged.neoforge.event.entity.living.DamageBlockEvent ev;
++            if (p_21017_ > 0.0F && (ev = this.damageContainer.setBlockedDamage(net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainer, this.isDamageSourceBlocked(p_21016_)))).getBlocked()) {
```
Is there a way to de-inline this statement?  This level of complexity in this single line makes it hard to read (which means it may be hard to maintain)

Caltinor replied:
I will break the update of the damage container to happen after the condition.  The rest should still remain in the condition.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -183,6 +203,14 @@
+             if (entity instanceof ServerPlayer) {
+                 CriteriaTriggers.PLAYER_HURT_ENTITY.trigger((ServerPlayer)entity, this, p_21016_, f, p_21017_, flag);
+             }
+-
```
Should adjust this here to avoid the whitespace removal patch.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -318,9 +346,20 @@
+     protected void doHurtEquipment(DamageSource p_330843_, float p_330394_, EquipmentSlot... p_331314_) {
+         if (!(p_330394_ <= 0.0F)) {
+             int i = (int)Math.max(1.0F, p_330394_ / 4.0F);
+-
```
Same as the prior comment: re-add the empty line so the patch is a pure addition

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -318,9 +346,20 @@
+         if (!(p_330394_ <= 0.0F)) {
+             int i = (int)Math.max(1.0F, p_330394_ / 4.0F);
+-
++            net.neoforged.neoforge.common.CommonHooks.onArmorHurt(p_330843_, p_331314_, p_330394_, this); p_331314_ = new EquipmentSlot[0];
```
Do not nest multiple statements in a single line like this.  The `p_331314_ = new EquipmentSlot[0];` statement should be on its own line, with an explicit comment to note that it is for the purpose of invalidating the following for loop.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -330,23 +369,48 @@
-             p_21241_ = this.getDamageAfterMagicAbsorb(p_21240_, p_21241_);
-             float f1 = Math.max(p_21241_ - this.getAbsorptionAmount(), 0.0F);
-@@ -1708,6 +_,7 @@
+-        if (!this.isInvulnerableTo(p_21240_)) {
```
Why was the invulnerable check removed here?  It is possible to invoke `actuallyHurt` without going through `hurt`, so the check should remain here.

Caltinor replied:
`actuallyHurt` is a protected method and is only ever invoked within `hurt` which is where we first check the invulnerability.  The removal was to remain consistent throughout the pipeline and ensure invulnerability check happens at the very beginning, isn't repeatedly invoked, and properly treats `actuallyHurt` as a final application of damage for the purpose of clarity.

Caltinor replied:
after discord discussion this call has been restored

Shadows-of-Fire commented on patches/net/minecraft/world/entity/player/Player.java.patch:
```diff
@@ -117,25 +117,35 @@
+                 }
+ 
+-                return p_36155_ == 0.0F ? false : super.hurt(p_36154_, p_36155_);
++                if (this.damageContainer.getNewDamage() == 0.0F) {this.damageContainer = null; return false;} else return super.hurt(p_36154_, this.damageContainer.getNewDamage());
```
This should not be a single line.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/player/Player.java.patch:
```diff
@@ -160,20 +170,45 @@
-             float f = p_36313_ - f1;
-             if (f > 0.0F && f < 3.4028235E37F) {
-                 this.awardStat(Stats.DAMAGE_ABSORBED, Math.round(f * 10.0F));
+-        if (!this.isInvulnerableTo(p_36312_)) {
```
Same comment about the removal of `isInvulnerableTo` in the earlier `actuallyHurt` patch

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -987,8 +1008,8 @@ public static void onEntityEnterSection(Entity entity, long packedOldPos, long p
-    public static ShieldBlockEvent onShieldBlock(LivingEntity blocker, DamageSource source, float blocked) {
-        ShieldBlockEvent e = new ShieldBlockEvent(blocker, source, blocked);
+    public static DamageBlockEvent onDamageBlock(LivingEntity blocker, DamageContainer container, boolean originalBlocked) {
+        DamageBlockEvent e = new DamageBlockEvent(blocker, container, originalBlocked);
```
Each one of these methods needs a javadoc detailing their intended firing site, parameters, and return values.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,304 @@
+ * undo the preceding effects</li>
+ * <li>{@link net.neoforged.neoforge.event.entity.living.DamageTakenEvent DamageTakenEvent} fires and provides
+ * an immutable perspective of what the entire sequence ended with. </li>
+ * </ol>
```
Javadoc references should typically use imported names (instead of FQNs) to reduce verboseness outside of an IDE.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,304 @@
+        /** Damage reduced from the effects of armor */
+        ARMOR,
+        /** Damage reduced from enchantments on armor */
+        ENCHANT,
+        /** Damage reduced from active mob effects */
+        MOBEFFECT,
```
Nit: these should probably be `ENCHANTMENTS` and `MOB_EFFECTS` for consistency with the other type names
Ideally you could write (in a javadoc) `Damage reduced by {@link Reduction.ENCHANTMENTS}` and have that sentence be grammatically correct.

Caltinor replied:
rename refactor applied.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,304 @@
+     *
+     * @return The amount of damage reduced by armor after vanilla armor reductions and added modifiers
+     */
+    float getArmorReduction();
+
+    /**
+     * This provides a post-reduction value for enchantment reduction and modifiers. This will always return zero
+     * before {@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent} and will consume all
+     * modifiers prior to the event.
+     *
+     * @return the amount of damage reduced by enchantments after vanilla enchantment reductions and added modifiers
+     */
+    float getEnchantReduction();
+
+    /**
+     * This provides a post-reduction value for mob effect reduction and modifiers. This will always return zero
+     * before {@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent} and will consume all
+     * modifiers prior to the event.
+     *
+     * @return The amount of damage reduced by mob effects after vanilla mob effect reductions and added modifiers
+     */
+    float getMobEffectReduction();
+
+    /**
+     * This provides a post-reduction value for absorption consumption and modifiers. This will always return zero
+     * before {@link net.neoforged.neoforge.event.entity.living.IncomingDamageEvent} and will consume all
+     * modifiers prior to the event.
+     *
+     * @return The amount of absorption consumed after vanilla absorption consumption and added modifiers
+     */
+    float getAbsorption();
```
Given the `Reduction` enum, should these be collapsed into one method with a `Reduction` parameter?

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,304 @@
+    float getAbsorption();
+
+    @ApiStatus.Internal
+    public class InternalDamageContainer implements DamageContainer {
```
The internal implementation should be in a standalone class, not an inner class of the public API surface.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/EntityInvulnerablityCheckEvent.java:
```diff
@@ -0,0 +1,57 @@
+ * Fired when {@link Entity#hurt(DamageSource, float)} is invoked and determines if
+ * downstream hurt logic should apply. This event is fired in {@link Entity#isInvulnerableTo(DamageSource)}
+ * <p>
+ * This event cannot be cancelled.
+ * <p>
+ * This event does not have a result.
+ * <p>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}.
```
These lines are not necessary, they are implicit from the event's class declaration (old events have them due to the old annotation system).

In their place I find it is most applicable to put a line noting which logical side the event will be fired on.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/ArmorHurtEvent.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Fired when a {@link LivingEntity}'s armor is dealt damage in {@link LivingEntity#doHurtEquipment(DamageSource, float, EquipmentSlot...) doHurtEquipment}.
+ * <p>
+ * This event is {@link ICancellableEvent cancelable}. Cancelling this event will prevent all damage to armor.
+ * <p>
+ * This event does not have a result.
+ * <p>
+ * This event is fired on the {@link NeoForge#EVENT_BUS}
```
Similar comment as `EntityInvulnerablityCheckEvent` - though since this event is `ICancellableEvent`, the documentation on what cancelling the event does should be placed in an override of `ICancellableEvent#setCanceled`.

Shadows-of-Fire reviewed (changes requested):
I don't have time to finish all the files atm (I got about halfway through) but I'll submit this review set thus far.  Lots of little changes that need to be made for sanity.

One thing I think needs to be looked at on a broad level is the event naming.  The sequence doesn't have a coherent name set, and we have two clashing names that (at a glance) would appear to be the "initiating" event - `EntityPreDamageEvent` and `IncomingDamageEvent`.  That might be better discussed on discord though.

Caltinor committed:
Address comments

Caltinor committed:
fix formatting issues.

Caltinor committed:
restore actuallyHurt invul checks

Caltinor committed:
collapse reduction methods

Caltinor committed:
remove FQNs from javadocs

Caltinor committed:
fix missed formatting issue

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageBlockEvent.java:
```diff
@@ -0,0 +1,117 @@
+ * true, The shield item stack "should" be available from {@link LivingEntity#getUseItem()} at least
+ * for players.</h4>
+ */
+public class DamageBlockEvent extends DamageSequenceEvent implements ICancellableEvent {
```
~~To match `DamageTakenEvent`, this should likely be `DamageBlockedEvent`.  `DamageBlockEvent` could be confused with an operation that happens to a `Block`, rather than when something blocks an instance of damage.~~

Edit: See naming discussion on discord

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -9,6 +9,18 @@
+     protected float appliedScale = 1.0F;
++    /**This field stores information about damage dealt to this entity.
++     * The field is instantiated via {@link #hurt(DamageSource, float)} after
++     * invulnerability checks, and is nulllified before the method's return.*/
```
Javadoc opening and closing lines (`/**` and `*/`) should be on their own line.

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -9,6 +9,18 @@
++    /**This field stores information about damage dealt to this entity.
++     * The field is instantiated via {@link #hurt(DamageSource, float)} after
++     * invulnerability checks, and is nulllified before the method's return.*/
++    @org.jetbrains.annotations.Nullable
```
Use javax `@Nullable` since it's already imported into the patch

Shadows-of-Fire commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -152,36 +164,57 @@
++            net.neoforged.neoforge.event.entity.living.DamageBlockEvent ev;
++            if (p_21017_ > 0.0F && (ev = net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainer, this.isDamageSourceBlocked(p_21016_))).getBlocked()) {
++                this.damageContainer.setBlockedDamage(ev);
++                if(ev.shieldDamage() > 0) this.hurtCurrentlyUsedShield(ev.shieldDamage());
```
Use braces on conditional blocks
```java
if (ev.shieldDamage() > 0) {
    this.hurtCurrentlyUsedShield(ev.shieldDamage());
}
```

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/IncomingDamageEvent.java:
```diff
@@ -10,6 +10,7 @@
+import net.neoforged.neoforge.common.damagesource.DamageContainer;
 
 /**
  * LivingHurtEvent is fired when an Entity is set to be hurt. <br>
```
Stale javadoc

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/EntityPreDamageEvent.java:
```diff
@@ -11,6 +11,7 @@
+import net.neoforged.neoforge.common.damagesource.DamageContainer;
 
 /**
  * LivingAttackEvent is fired when a living Entity is attacked. <br>
```
Stale javadoc

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageTakenEvent.java:
```diff
@@ -9,6 +9,7 @@
+import net.neoforged.neoforge.common.damagesource.DamageContainer;
 
 /**
  * LivingDamageEvent is fired just before damage is applied to entity.<br>
```
Stale javadoc

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageBlockEvent.java:
```diff
@@ -0,0 +1,117 @@
+import net.neoforged.neoforge.common.damagesource.DamageContainer;
+
+/**
+ * The ShieldBlockEvent is fired when an entity is hurt and vanilla checks if the entity is attempting
```
Stale javadoc

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        @Override
+        public void addModifier(Reduction type, BiFunction<DamageContainer, Float, Float> function) {}
+
+        @Override
+        public void setNewDamage(float damage) {}
+
+        @Override
+        public void setPostAttackInvulnerabilityTicks(int ticks) {}
```
These setter methods should throw when invoked on the immutable container

Caltinor committed:
Update javadocs and formatting

Caltinor committed:
refactor event names

Caltinor committed:
fix merge error

Caltinor committed:
event refactor in tests fix

Caltinor committed:
update javadoc and fix name typo

Caltinor committed:
update damage container as Stack

Tslat commented on src/main/java/net/neoforged/neoforge/event/entity/living/LivingDamageEvent.java:
```diff
@@ -1,54 +1,74 @@
-    public void setAmount(float amount) {
-        this.amount = amount;
+    /**
+     * DamageTakenEvent is fired just before health is modified on the entity.<br>
```
This javadoc is out of date

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -350,23 +406,43 @@
+                 this.setAbsorptionAmount(this.getAbsorptionAmount() - f1);
+                 this.gameEvent(GameEvent.ENTITY_DAMAGE);
++                this.onDamageTaken(this.damageContainer.getLast());
++                net.neoforged.neoforge.common.CommonHooks.onLivingDamagePost(this, this.damageContainer.getLast());
```
I know we already discussed this, but I just remembered why I specifically wanted the `Post` event __outside__ of these braces in the first place.

Because this handling is done inside of `actuallyHurt`, the remainder of the handling that happens in `hurt` does not take this value into account in any way, and proceeds based on the value that was initially passed to hurt (why, Mojang?)
This functionally means that all the damage modifications after the shield block __do not__ have an effect on the resultant triggering of knockback, hurt marking, etc.

TL;DR: The attack is still doing things after this, even if the value is 0. This event should be called after the closing brace of this section instead, so that callers can still hook successful attacks, even if the resultant "damage" is 0

I'm not sure if the same should apply to `onDamageTaken`, I'm open to debate on that since I can see valid justifications either way

Tslat commented on src/main/java/net/neoforged/neoforge/event/entity/living/LivingIncomingDamageEvent.java:
```diff
@@ -0,0 +1,41 @@
+ * 
+ * @see DamageSequenceEvent
+ **/
+public class LivingIncomingDamageEvent extends DamageSequenceEvent implements ICancellableEvent {
```
Should we add helper methods to this event to allow people to directly add modifiers to the damage container?

This provides 3 benefits:
1. It makes it apparent to casual listeners that modifying the intervening damage stages is even _possible_ now
2. It makes it much clearer than this event is the place those modifications should be registered
3. It makes it a little nicer to actually add modifiers, since you don't need to get the container first

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -152,36 +166,59 @@
-         } else if (this.level().isClientSide) {
-@@ -1136,14 +_,17 @@
+         } else {
++            this.damageContainer.add(new net.neoforged.neoforge.common.damagesource.InternalDamageContainer(p_21016_, p_21017_));
```
Should this not be `Stack#push`?

`add` will bypass the counter

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -9,6 +9,20 @@
++     * invulnerability checks, and is nulllified before the method's return.
++     * */
++    @Nullable
++    protected java.util.Stack<net.neoforged.neoforge.common.damagesource.InternalDamageContainer> damageContainer = new java.util.Stack<>();
```
With the change to a `Stack`, should this be renamed to `damageContainers`

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -152,36 +166,59 @@
+             }
+ 
+             this.noActionTime = 0;
++            p_21017_ = this.damageContainer.getLast().getNewDamage(); //Neo: enforce damage container as source of truth for damage amount
```
`Stack#peek`?

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -152,36 +166,59 @@
-+                if(!ev.isCanceled()) {
-+                if(ev.shieldTakesDamage()) this.hurtCurrentlyUsedShield(p_21017_);
++            net.neoforged.neoforge.event.entity.living.LivingShieldBlockEvent ev;
++            if (p_21017_ > 0.0F && (ev = net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainer.getLast(), this.isDamageSourceBlocked(p_21016_))).getBlocked()) {
```
`Stack#peek`?

Tslat commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -152,36 +166,59 @@
-+                if(ev.shieldTakesDamage()) this.hurtCurrentlyUsedShield(p_21017_);
++            net.neoforged.neoforge.event.entity.living.LivingShieldBlockEvent ev;
++            if (p_21017_ > 0.0F && (ev = net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainer.getLast(), this.isDamageSourceBlocked(p_21016_))).getBlocked()) {
++                this.damageContainer.getLast().setBlockedDamage(ev);
```
`Stack#peek`?

I'm not gonna do any more of these lol, just do a sweep if you decide to change the call

Caltinor committed:
update Stack method calls

Caltinor committed:
update javadocs and add helper methods

Caltinor committed:
update javadoc

Caltinor committed:
rebase patches

Caltinor committed:
more rebase fixes

Caltinor committed:
final rebase fixes

Caltinor committed:
spotless cleanup

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Matyrobbrt commented on patches/net/minecraft/world/entity/Entity.java.patch:
```diff
@@ -292,6 +292,22 @@
+-            || p_20122_.is(DamageTypeTags.IS_FALL) && this.getType().is(EntityTypeTags.FALL_DAMAGE_IMMUNE);
++        boolean isVanillaInvulnerable = this.isRemoved()
++                || this.invulnerable && !p_20122_.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !p_20122_.isCreativePlayer()
++                || p_20122_.is(DamageTypeTags.IS_FIRE) && this.fireImmune()
```
You're needlessly indenting and causing patches here.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -238,12 +240,48 @@ public static LivingChangeTargetEvent onLivingChangeTarget(LivingEntity entity,
+     * @return if the event is cancelled and no damage will be applied to the entity
+     */
+    public static boolean onPlayerIncomingDamage(LivingEntity entity, DamageContainer container) {
+        return !NeoForge.EVENT_BUS.post(new LivingIncomingDamageEvent(entity, container)).isCanceled();
```
This check is inverted?

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,161 @@
+ *
+ *
+ */
+public interface DamageContainer {
```
Since you're dealing with both mutable and immutable containers, I feel like you should have a super `DamageContainer` and then a `MutableDamageContainer` to indicate the differenece more clearly.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,161 @@
+     *                  value for the reduction. These are always executed in insertion order. if sequence
+     *                  matters, use {@link net.neoforged.bus.api.EventPriority} to order your function.
+     */
+    void addModifier(Reduction type, BiFunction<DamageContainer, Float, Float> operation);
```
Please use a specialised functional interface that's properly documented.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/InternalDamageContainer.java:
```diff
@@ -0,0 +1,105 @@
+
+@ApiStatus.Internal
+public class InternalDamageContainer implements DamageContainer {
+    private final EnumMap<Reduction, List<BiFunction<DamageContainer, Float, Float>>> reductionMap = new EnumMap<>(Reduction.class);
```
`reductionFunctions` seems like a better name

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/InternalDamageContainer.java:
```diff
@@ -0,0 +1,105 @@
+    private final float originalDamage;
+    private final DamageSource source;
+    private float newDamage;
+    EnumMap<Reduction, Float> reductions = new EnumMap<>(Reduction.class);
```
Why is this not final and private?

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/entity/EntityInvulnerabilityCheckEvent.java:
```diff
@@ -0,0 +1,55 @@
+        this.isInvulnerable = isInvulnerable;
+    }
+
+    /** @return the current invulnerability state */
+    public boolean isInvulnerable() {
+        return isInvulnerable;
+    }
+
+    /** @return an immutable reference to the damage source being applied to this entity */
+    public DamageSource getSource() {
+        return source;
+    }
+
+    /** @return the invulnerability status passed into the event by vanilla */
+    public boolean getOriginalInvulnerability() {
+        return originallyInvulnerable;
+    }
```
These docs can be inline returns.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/entity/living/ArmorHurtEvent.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private final EnumMap<EquipmentSlot, ArmorEntry> armorEntries;
```
The map itself can be immutable.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/entity/living/ArmorHurtEvent.java:
```diff
@@ -0,0 +1,74 @@
+     * @param damage the new amount to hurt the armor. Values below zero will be set to zero.
+     */
+    public void setNewDamage(EquipmentSlot slot, float damage) {
+        if (this.armorEntries.containsKey(slot)) this.armorEntries.get(slot).newDamage = damage;
```
You can use a `get` and null check.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/entity/living/DamageSequenceEvent.java:
```diff
@@ -0,0 +1,51 @@
+ * can implement this interface if their corresponding hooks effectively terminate
+ * the damage sequence.
+ * <br>
+ * <H3>The Damage Sequence</H1>
```
Er... `h3` closed by `h1`? Also, tags should be lowercase.

Caltinor committed:
addressing review

- remove DamageSequenceEvent and refactor events to extend LivingEvent only with DamageContainers as applicable
- flatten DamageContainer to one class instance

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
fix typo and clean javadocs

Caltinor committed:
Fix refactor not applied to patch

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,181 @@
+ * DamageContainer encapsulates aspects of the entity damage sequence so that
+ * relevant context related to damage dealt is accessible throughout the entire
+ * sequence.
+ * <h4>Note: certain values will be defaults until the stage in the sequence when they are set.</h4>
```
You can put this in a paragraph, it's not a header.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,181 @@
+ * <h4>Note: certain values will be defaults until the stage in the sequence when they are set.</h4>
+ * <h3>The Damage Sequence</h3>
+ * <ol>
+ * <li>{@link LivingEntity#hurt} is invoked on the recipient from the source of
```
You either end all list entries with a dot, or not.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,181 @@
+        ENCHANTMENTS,
+        /** Damage reduced from active mob effects */
+        MOB_EFFECTS,
+        /** Damage absorbed by absorption. */
```
Same, all dots or no dot.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/damagesource/DamageContainer.java:
```diff
@@ -0,0 +1,181 @@
+     * @param type              The reduction type your function will apply to
+     * @param reductionFunction takes the current reduction from vanilla and any preceding functions and returns a new
+     *                          value for the reduction. These are always executed in insertion order. if sequence
+     *                          matters, use {@link net.neoforged.bus.api.EventPriority} to order your function.
+     */
+
+    public void addModifier(Reduction type, IReductionFunction reductionFunction) {
+        this.reductionFunctions.computeIfAbsent(type, a -> new ArrayList<>()).add(reductionFunction);
+    }
+
+    /**
+     * @return The damage blocked during the {@link LivingShieldBlockEvent}
+     */
+    public float getBlockedDamage() {
+        return blockedDamage;
+    }
+
+    /**
+     * @return The durability applied to the applicable shield after {@link LivingShieldBlockEvent}
+     *         returned a successful block
+     */
+    public float getShieldDamage() {
+        return shieldDamage;
+    }
+
+    /**
+     * Explicitly sets the invulnerability ticks after the damage has been applied.
+     *
+     * @param ticks Ticks of invulnerability after this damage sequence
+     */
+    public void setPostAttackInvulnerabilityTicks(int ticks) {
+        this.invulnerabilityTicksAfterAttack = ticks;
+    }
+
+    /**
+     * @return The number of ticks this entity will be invulnerable after damage is applied.
+     */
+    public int getPostAttackInvulnerabilityTicks() {
+        return invulnerabilityTicksAfterAttack;
+    }
```
I'd make these @return tags inline {@return}.
Remember not to end them with a dot and to start them with lowercase.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/event/entity/living/LivingIncomingDamageEvent.java:
```diff
@@ -0,0 +1,85 @@
+ * <br>
+ * This event is fired via the {@link CommonHooks#onEntityIncomingDamage(LivingEntity, DamageContainer)}.<br>
+ * <br>
+ * For more information on the damage sequence
+ * 
+ * @see DamageContainer
```
Err, this isn't exactly a proper javadoc.
```suggestion
 * @see DamageContainer for more information on the damage sequence
```

Matyrobbrt commented on tests/src/main/java/net/neoforged/neoforge/debug/entity/EntityEventTests.java:
```diff
@@ -75,6 +80,24 @@ static void entityAttributeModificationEvent(final DynamicTest test, final Regis
+    @TestHolder(description = "Tests if EntityInvulnerabilityCheckEvent prevents damage when modified.")
+    static void entityInvulnerabilityCheckEvent(final DynamicTest test, final RegistrationHelper reg) {
+        test.eventListeners().forge().addListener((final EntityInvulnerabilityCheckEvent event) -> {
+            if (event.getEntity() instanceof GameTestPlayer entity)
```
"wildcard" tests like these are dangerous, please add a marker to distinguish the players. Maybe change its name.

Matyrobbrt commented on tests/src/main/java/net/neoforged/neoforge/debug/entity/EntityEventTests.java:
```diff
@@ -75,6 +80,24 @@ static void entityAttributeModificationEvent(final DynamicTest test, final Regis
+            DamageSource source = new DamageSource(helper.getLevel().registryAccess().registryOrThrow(Registries.DAMAGE_TYPE).getHolderOrThrow(DamageTypes.MOB_ATTACK));
+            helper.startSequence(() -> helper.makeTickingMockServerPlayerInLevel(GameType.SURVIVAL))
+                    .thenExecute(player -> player.setInvulnerable(true))
+                    .thenWaitUntil(player -> helper.assertTrue(!player.isInvulnerableTo(source), "Player Invulnerability not bypassed."))
```
So... `assertFalse`

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
additional code cleanup

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

TelepathicGrunt commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -330,9 +387,20 @@
+             int i = (int)Math.max(1.0F, p_330394_ / 4.0F);
+ 
++            net.neoforged.neoforge.common.CommonHooks.onArmorHurt(p_330843_, p_331314_, p_330394_, this);
++            p_331314_ = new EquipmentSlot[0]; //Neo: invalidates the loop.  armor damage happens in common hook
```
Couldn't this be a `if (true) return;` so it doesn't need to make the array at all? Intellij won't get angry like it does with just `return;`

Caltinor replied:
Wouldn't that just cause your IDE to yell at you for unreachable code?  I don't see how that's functionally better.

TelepathicGrunt replied:
It doesn't yell at me. I guess the current code is fine. Was thinking we could just early exit 
![image](https://github.com/neoforged/NeoForge/assets/40846040/6089df17-38c8-46d8-951a-5aa5407f4153)

Shadows-of-Fire replied:
Generally speaking `if (true/false) { ...` are the way that we null out code blocks we want to avoid.

The IDE will produce a warning for the code below `if (true) return;` (similar for code within an `if (false) {` block), but we don't care about warnings in the patched vanilla source, there's ~15k of them already.

Since there is no additional code below this line that needs to be run, using if (true) return;` is more expressive of the intent here (to cancel the rest of the method), so let's use that.

Caltinor replied:
changed.

TelepathicGrunt commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -330,9 +387,20 @@
++                    this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.MOB_EFFECTS, f2);
                      if (this instanceof ServerPlayer) {
 -                        ((ServerPlayer)this).awardStat(Stats.DAMAGE_RESISTED, Math.round(f2 * 10.0F));
 +                        ((ServerPlayer)this).awardStat(Stats.CUSTOM.get(Stats.DAMAGE_RESISTED), Math.round(f2 * 10.0F));
```
Why the change here? Isn't the vanilla code equivalent?
![image](https://github.com/neoforged/NeoForge/assets/40846040/cdbbc06d-ad43-4c51-a494-087b4cc8efa8)

Caltinor replied:
that's not my change.  That's from a previous PR but since it falls in the same hunk as my change, it shows up.

TelepathicGrunt replied:
huh... I'll have to look at this closer...

TelepathicGrunt commented on src/main/java/net/neoforged/neoforge/common/extensions/ILivingEntityExtension.java:
```diff
@@ -64,4 +66,15 @@ default boolean canDrownInFluidType(FluidType type) {
+    /**
+     * Executes in {@link LivingEntity#hurt(DamageSource, float)} after all damage and
+     * effects have applied. Overriding this method is preferred over overriding the
+     * hurt method in custom entities where special behavior is desired after vanilla
```
So in my case where I override hurt in my entity to cap the max damage to never more than 3, I would need to stop overriding this in my common code somehow and override onDamageTaken instead by mixin into my common entity class and overriding this method instead?
https://github.com/TelepathicGrunt/Bumblezone/blob/2e853574f4ca68e1a644de4f6c8daef13108794f/common/src/main/java/com/telepathicgrunt/the_bumblezone/entities/living/CosmicCrystalEntity.java#L1406

What is the issues that can happen if I refuse to stop overriding hurt?

Caltinor replied:
I would suggest calling `super.hurt` at the end of your override and including in your override only the parts that differ.  that said, looking at your override, you may want to just include the logic for adding to the damageContainer stack, firing the first event, and then popping the stack.  Whichever is cleaner, but both are options.

overriding `onDamageTaken` would be too late in the damage sequence to be of use for limiting the damage.  That callback is there for behavior that happens at the time of damage received but doesn't influence the damage sequence, eg. emitting particles on the entity based on the damage source and value as a custom mob feature.

TelepathicGrunt commented on src/main/java/net/neoforged/neoforge/event/entity/EntityInvulnerabilityCheckEvent.java:
```diff
@@ -0,0 +1,55 @@
+ * {@link Entity#isInvulnerableTo(DamageSource)}
+ * <br>
+ * <b>Note: This event may be unable to change the invulnerable status of some entities
+ * that override isInvulnerableTo against certain damage sources</b>
```
I noticed that Ghast fireballs that users hits reflected back will cause this event to not fire so you can't make Ghasts immune to reflected fireballs. Is there any way we can make this event a bit more useful? Or is the fear that patching every entity that overrides isInvulnerableTo to fire the event first, something that is too much to do?

Wondering if it could be an asm check that wraps the entirety of all isInverulerableTo methods into the onEntityInvulnerabilityCheck method. But then the event could fire multiple times for a single entity that calls super...

Annoying.

Caltinor replied:
So we aren't really concerned about multiple firing of the invulnerability check.  That came up in discussion.  If there is a clean way to capture those invulnerability checks with overrides using a fancy ASM, i welcome the input.  

Ultimately you are right about why we didn't inject into these entities.  There are 3 entities with special cases that result in the invulnerability check not firing: Breezes, Ghasts, and the Warden.  Anyone who needs to circumvent the vanilla behavior for these cases can address that themselves, but for the sake of keeping this feature clean, they were left out.

Shadows-of-Fire replied:
That may be an avenue to explore in the future, but is currently blocked by https://github.com/neoforged/NeoForge/issues/1168

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -238,12 +241,32 @@ public static LivingChangeTargetEvent onLivingChangeTarget(LivingEntity entity,
+     * @param isInvul whether this entity is invulnerable according to preceding/vanilla logic
+     * @return if this entity is invulnerable
+     */
+    public static boolean onEntityInvulnerabilityCheck(Entity entity, DamageSource source, boolean isInvul) {
```
rename this to `isEntityInvulnerableTo`

Caltinor committed:
IDE friendly code replacement

Caltinor committed:
Merge branch '1.20.x' of https://github.com/Caltinor/NeoForge into 1.20.x

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -238,12 +241,32 @@ public static LivingChangeTargetEvent onLivingChangeTarget(LivingEntity entity,
+    /**
+     * Called after invulnerability checks in {@link LivingEntity#hurt(DamageSource, float)},
+     * this method creates and posts the first event in the LivingEntity damage sequence,
+     * {@link LivingIncomingDamageEvent}, FOR NON-PLAYER entities.
```
> FOR NON-PLAYER entities

Is this still the case after we removed the player-specific overrides?

Caltinor replied:
It is not.  Fixed.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/common/CommonHooks.java:
```diff
@@ -256,14 +279,58 @@ public static boolean onLivingUseTotem(LivingEntity entity, DamageSource damageS
+     * @param container the container object holding the final values of the damage pipeline while they are
+     *                  still mutable
+     * @return the current damage value to be applied to the entity's health
+     *
```
extra whitespace line (I think spotless should catch this, but I guess it doesn't)

Caltinor replied:
that's not enough words to new-line.  i'll inline it.

Caltinor committed:
javadoc tweaks

Caltinor committed:
rename invul check in CommonHooks

Caltinor committed:
Merge branch '1.21.x' into 1.20.x

Matyrobbrt committed:
Merge branch '1.21.x' into 1.20.x
